### PR TITLE
Bugfixing kubeconfig decoding

### PR DESCRIPTION
**Pull request by ghost**
## Changes introduced with this PR

This PR adds several bugfixes to the Kubeconfig parsing. Sadly, there has not been time to add tests for all edge cases yet.

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: kubeconfig_plugin.py [hunk 1]
```diff
@@ -111,66 +133,116 @@ class ErrorOutput:
     id="kubeconfig",
     name="kubeconfig plugin",
     description=(
-        "Inputs a kubeconfig, parses it and extracts the kubernetes cluster"
-        " details "
+            "Inputs a kubeconfig, parses it and extracts the kubernetes cluster"
```
Comment (sandrobonazzola) [inline kubeconfig_plugin.py]: any reason for this additional indentation?

File: tests/test_client_cert.yaml [hunk 2]
```diff
@@ -13,13 +13,13 @@ kubeconfig: |
...
+      client-key-data: LS0tLS1CRUdJTiBQUklWQVRFIEtFWS0tLS0tCk1JSUJWQUlCQURBTkJna3Foa2lHOXcwQkFRRUZBQVNDQVQ0d2dnRTZBZ0VBQWtFQXJyODlmMmtnZ1NPL3lhQ0IKNkV3SVFlVDZacHRCb1gwWnZDTUkrRHBrQ3dxT1M1ZndSYmoxbkVpUG5MYnpERGdNVThLQ1BBTWhJN0pwWVJsSApuaXB4V3dJREFRQUJBa0J5YnUveDBNRWxjR2kydS9KMlVkd1Njc1Y3amU1VHQxMno4Mmw3VEptWkZGSjhSTG1jCnJoMDBHdmViNFZwR2hkMStjM2xaYk8xbUlUNnYzdkhNOUEwaEFpRUExNEVXNmIrOTlYWXphNys1dXdJRHVpTSsKQnozcGtLKzl0bGZWWEU3SnlLc0NJUURQbFlKNXh0YnVUK1Z2QjNYT2REL1ZXaUVxRW12RTNmbFYwNDE3UnFoYQpFUUlnYnl4d05wd3RFZ0V0Vzh1bnRCckE4M2lVMmtXTlJZL3o3YXA0TGt1Uyswc0NJR2UyRSswUm1mcVFzbGxwCmljTXZNMkU5MllueWtDTlluNlR3d0NRU0pqUnhBaUVBbzlNbWFWbEs3WWRoU01QbzUydUpZemQ5TVFaSnFocSsKbEIxWkdEeC9BUkU9Ci0tLS0tRU5EIFBSSVZBVEUgS0VZLS0tLS0K
```
Comment (sandrobonazzola) [inline tests/test_client_cert.yaml]: missing new line at end of file

**Review by sandrobonazzola — CHANGES_REQUESTED** [review]
A few inline comments. Also https://github.com/arcalot/arcaflow-plugin-kubeconfig/actions/runs/3272042293/jobs/5382567090#step:4:27

File: kubeconfig_plugin.py (hunk 1 above)
Comment (ghost) [reply kubeconfig_plugin.py]: Only reformatting.

File: kubeconfig_plugin.py [hunk 3]
```diff
@@ -117,57 +138,106 @@ class ErrorOutput:
...
-
-        if (
-            "client_key_data" not in output
-            and "client_certificate_data" not in output
-            and "token" not in output
-        ):
+        try:
+            context = None
+            for ctx in kubeconfig["contexts"]:
+                if ctx["name"] == current_context:
```
Comment (engelmi) [inline kubeconfig_plugin.py:159]: Just to make sure: `current_context` from `kubeconfig["current-context"]` is a `string` and not a `dict`, right?

File: kubeconfig_plugin.py [hunk 4]
```diff
@@ -117,57 +138,106 @@ class ErrorOutput:
...
+                    "Failed to find a user named {} in the kubeconfig file.".format(current_user))
+        except Exception as e:
+            return "error", ErrorOutput(
+                "Failed to find a user named {} in the kubeconfig file: {}".format(current_user, e.__str__()))
 
-        return "success", kubeconfig_output_schema.unserialize(output)
-    except Exception:
-        return "error", ErrorOutput(1, "Failure in parsing kubeconfig:")
+        output = SuccessOutput(
+            Connection(cluster["server"]),
```
Comment (Harshith-umesh) [inline kubeconfig_plugin.py]: can we first assign cluster["server"] as host and then use that, it'll reduce confusion in the future

File: kubeconfig_plugin.py [hunk 5]
```diff
@@ -117,57 +138,106 @@ class ErrorOutput:
...
+                                                            "certificate-authority-data"
+                                                        ]).decode('ascii')
+        except KeyError:
+            pass
+        try:
+            output.connection.cert = base64.b64decode(user[
+                                                          "client-certificate-data"
+                                                      ]).decode('ascii')
+        except KeyError:
+            pass
```
Comment (engelmi) [inline kubeconfig_plugin.py]: Could you wrap thos in a dedicated function? 
```python
def getFromDict(d -> dict, key -> str) -> str:
   try:
      return d[key]
   except KeyError:
      pass
```

**Review by dustinblack — APPROVED** [review]
I only did a cursory code review due to time constraints, but I did do a functional test, and my workflow was able to read the kubeconfig and then deploy a pod into a kubernetes cluster.

File: kubeconfig_plugin.py (hunk 3 above)
Comment (ghost) [reply kubeconfig_plugin.py:159]: Correct.

File: kubeconfig_plugin.py (hunk 5 above)
Comment (ghost) [reply kubeconfig_plugin.py]: That wouldn't work, because if you pass a `None` to b64decode, it won't work.

File: requirements.txt [hunk 6]
```diff
@@ -1,28 +1,3 @@
 arcaflow-plugin-sdk==0.8.0; python_version >= "3.9"
```
Comment (dustinblack) [inline requirements.txt]: Bump this to 0.9.0

**Review by dustinblack — CHANGES_REQUESTED** [review]

File: kubeconfig_plugin.py (hunk 5 above)
Comment (portante) [reply kubeconfig_plugin.py]: Why can't one use:
```
d.get(key, "")
```

File: kubeconfig_plugin.py [hunk 7]
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
     id="kubeconfig",
     name="kubeconfig plugin",
     description=(
-        "Inputs a kubeconfig, parses it and extracts the kubernetes cluster"
-        " details "
+            "Inputs a kubeconfig, parses it and extracts the kubernetes cluster"
+            " details "
```
Comment (portante) [inline kubeconfig_plugin.py]: `s/ details / details/`

File: kubeconfig_plugin.py [hunk 8]
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
...
-            output["client_key_data"] = kcl._user.value["client-key-data"]
-        except Exception:
-            print(
-                "client certificate and key data missing, trying to extract"
-                " token"
-            )
+            current_context = kubeconfig["current-context"]
+        except KeyError:
+            return "error", ErrorOutput(
+                "The provided kubeconfig file does not have a current-context set. Please set a current context to use.")
```
Comment (portante) [inline kubeconfig_plugin.py]: Consider the following to reduce the error returns and lines of code a bit:

```
kind = kubeconfig.get("kind", "")
if kind != "Config":
    addendum = "" if kind else " (missing 'kind' field)"
    return "error", ErrorOutput(f"The provided file is not a kubeconfig file{addendum}")
current_context = kubeconfig.get("current-context", "")
if not current_context:
    return "error", ErrorOutput(
                "The provided kubeconfig file does not have a current-context set. Please set a current context to use.")
```

File: kubeconfig_plugin.py [hunk 9]
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
...
+            context = None
+            for ctx in kubeconfig["contexts"]:
+                if ctx["name"] == current_context:
+                    context = ctx["context"]
+            if context is None:
+                return "error", ErrorOutput(
+                    "Failed to find a context named {} in the kubeconfig file.".format(current_context))
+            current_cluster = context["cluster"]
+            current_user = context["user"]
+        except Exception as e:
```
Comment (portante) [inline kubeconfig_plugin.py]: Shouldn't this be `except KeyError as e`?

File: kubeconfig_plugin.py [hunk 10]
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
...
             return "error", ErrorOutput(
-                1, "Both client data and token missing, exiting!"
-            )
+                "Failed to find a context named {} in the kubeconfig file: {}".format(current_context, e.__str__()))
```
Comment (portante) [inline kubeconfig_plugin.py]: Consider:

```
return "error", ErrorOutput(f"Failed to find a context named {current_context} in the kubeconfig file: {e!r}")
```

File: kubeconfig_plugin.py [hunk 11]
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
...
+            cluster = None
+            for cl in kubeconfig["clusters"]:
+                if cl["name"] == current_cluster:
+                    cluster = cl["cluster"]
+            if cluster is None:
+                return "error", ErrorOutput(
+                    "Failed to find a cluster named {} in the kubeconfig file.".format(current_cluster))
+        except Exception as e:
+            return "error", ErrorOutput(
+                "Failed to find a cluster named {} in the kubeconfig file: {}".format(current_cluster, e.__str__()))
```
Comment (portante) [inline kubeconfig_plugin.py]: Consider a bug fix (need to prevent `KeyError` in the inner loop):

```
cluster = None
error_addendum = ""
try:
    clusters = kubeconfig["clusters"]
except Exception as e:
    error_addendum = f": {e}"
else:
    for cl in clusters:
        try:
            if cl["name"] == current_cluster:
                cluster = cl["cluster"]
        except KeyError:
            pass
if cluster is None:
    return "error", ErrorOutput(
        f"Failed to find a cluster named {current_cluster} in the kubeconfig file{error_addendum}"
    )
```

File: kubeconfig_plugin.py [hunk 12]
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
...
+            user = None
+            for u in kubeconfig["users"]:
+                if u["name"] == current_user:
+                    user = u["user"]
+            if user is None:
+                return "error", ErrorOutput(
+                    "Failed to find a user named {} in the kubeconfig file.".format(current_user))
+        except Exception as e:
+            return "error", ErrorOutput(
+                "Failed to find a user named {} in the kubeconfig file: {}".format(current_user, e.__str__()))
```
Comment (portante) [inline kubeconfig_plugin.py]: Consider a similar refactoring of lines 163 - 173 to what is proposed for changes in the comment for line 161.

File: kubeconfig_plugin.py [hunk 13]
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
...
 
-        return "success", kubeconfig_output_schema.unserialize(output)
-    except Exception:
-        return "error", ErrorOutput(1, "Failure in parsing kubeconfig:")
+        output = SuccessOutput(
+            Connection(host=cluster["server"]),
```
Comment (portante) [inline kubeconfig_plugin.py]: If `server` is not found in, shouldn't this report the `KeyError` without a huge stack that you'd get from the exception handling at line 210?  And perhaps consider moving this up to immediately follow lines 161?

File: kubeconfig_plugin.py (hunk 5 above)
Comment (portante) [reply kubeconfig_plugin.py]: If you can't have empty fields in the `Connection` data object then you need the try/except style.

**Review by portante — CHANGES_REQUESTED** [review]
One potential bug, consider some alternative approaches in the code.

**Review by portante — COMMENTED** [review]
Latest changes look fine.

File: kubeconfig_plugin.py (hunk 5 above)
Comment (jaredoconnell) [reply kubeconfig_plugin.py]: The format that the schema needs is `None` for all missing optional values. We can't use get for this, so the only ways for this to work is the current way, or a just-as-verbose `in` conditional.

Comment (jaredoconnell) [reply kubeconfig_plugin.py]: Fixed in next commit (in progress).

**Review by portante — COMMENTED** [review]
Looking good.

File: kubeconfig_plugin.py [hunk 14]
```diff
@@ -111,63 +109,92 @@ class ErrorOutput:
...
+            return "error", ErrorOutput(
+                "Failed to find a user named {} in the kubeconfig file: {}".format(current_user, e.__str__()))
+
+        if "server" not in cluster:
+            return "error", ErrorOutput(
+                "Failed to find server in cluster kubeconfig file {} cluster section".format(e.__str__()))
+        output = SuccessOutput(
+            Connection(host=cluster["server"]),
+        )
+	output.connection.cacert = base64_decode(cluster.get("certificate-authority-data", None))
```
Comment (dustinblack) [inline kubeconfig_plugin.py]: Leading white space here is a tab, not spaces, throwing an error when `test_kubeconfig_plugin.py` is run.

**Review by dustinblack — CHANGES_REQUESTED** [review]

**Review by sandrobonazzola — APPROVED** [review]

File: pyproject.toml [hunk 15]
```diff
@@ -2,16 +2,16 @@
...
-license = "Apache-2.0+GPL-2.0-only"
+authors = ["Arcalot Contributors"]
+license = "Apache-2.0"
 readme = "README.md"
 
 
 [tool.poetry.dependencies]
 python = "^3.9"
-arcaflow-plugin-sdk = "0.9.0"
+arcaflow-plugin-sdk = { git = "https://github.com/arcalot/arcaflow-plugin-sdk-python.git", branch="allow-scope-in-one-of" }
```
Comment (portante) [inline pyproject.toml]: Why are we not first merging the `allow-scope-in-one-of` branch to `main` in the `arcaflow-plugin-sdk-python` respository first?

**Review by portante — COMMENTED** [review]
Any consideration for the existing comments?

File: pyproject.toml (hunk 15 above)
Comment (ghost) [reply pyproject.toml]: Because it doesn't have adequate tests yet.

File: kubeconfig_plugin.py (hunk 13 above)
Comment (jaredoconnell) [reply kubeconfig_plugin.py]: Now handled with explicit check above.

File: kubeconfig_plugin.py (hunk 8 above)
Comment (jaredoconnell) [reply kubeconfig_plugin.py]: Code refactored.

**Comment by jaredoconnell** [comment]
I refactored the code and added a lot of test cases. The test cases cover the general cases, as well as each of the potential missing key situations.

**Review by mfleader — APPROVED** [review]

File: pyproject.toml (hunk 15 above)
Comment (dustinblack) [reply pyproject.toml]: Clearly we don't want to merge this PR until we are ready to replace this with a version number. We should convert this to a draft for now.

**Review by jaredoconnell — APPROVED** [review]
Confirmed to work with old SDK. I believe this is ready for merging. The extra checks make this a much safer option than the code in main.

**Review by sandrobonazzola — APPROVED** [review]
LGTM

**Review by dustinblack — APPROVED** [review]

File: kubeconfig_plugin.py [hunk 16]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
 
     try:
-        kubeconfig = yaml.safe_load(params.kubeconfig)
-        kcl = config.kube_config.KubeConfigLoader(kubeconfig)
+        try:
+            kubeconfig = yaml.safe_load(params.kubeconfig)
+        except Exception as e:
+            return "error", ErrorOutput(
+                "Exception occurred while loading YAML. Input is not valid YAML."
+                + " Exception: {}".format(e.__str__())
```
Comment (portante) [inline kubeconfig_plugin.py]: Wouldn't an f-string work here?

```
            return "error", ErrorOutput(
                "Exception occurred while loading YAML. Input is not valid YAML."
                f" Exception: {e}"
            )
```

File: kubeconfig_plugin.py [hunk 17]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
+        # Kubeconfig files have the kind set as Config
+        if "kind" in kubeconfig:
+            if kubeconfig["kind"] != "Config":
+                return "error", ErrorOutput(
+                    "The provided file is not a kubeconfig file"
+                )
+        else:
+            return "error", ErrorOutput(
+                "The provided file is not a kubeconfig file (missing 'kind' field)"
+            )
```
Comment (portante) [inline kubeconfig_plugin.py:138]: Consider a couple of alternatives if you are gonna be back in here:

```
        if "kind" not in kubeconfig:
            return "error", ErrorOutput(
                "The provided file is not a kubeconfig file (missing 'kind' field)"
            )
        if kubeconfig["kind"] != "Config":
            return "error", ErrorOutput(
                "The provided file is not a kubeconfig file"
            )
```

or a bit longer but you leverage a more pythonic way to do this where you only reference the dictionary once:

```
        try:
            kind = kubeconfig["kind"]
        except KeyError:
            return "error", ErrorOutput(
                "The provided file is not a kubeconfig file (missing 'kind' field)"
            )
        if kind != "Config":
            return "error", ErrorOutput(
                "The provided file is not a kubeconfig file"
            )
```

Comment (portante) [reply kubeconfig_plugin.py:138]: And this notion of checking a dictionary and then accessing the dictionary for the same element is littered below.  It is usually worth the effort to only access the dictionary once.

File: kubeconfig_plugin.py [hunk 18]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
-        except Exception:
-            print(
-                "client certificate and key data missing, trying to extract"
-                " token"
+        # Get the current context, then search for the values for that context in the
+        # context section
+        current_context = kubeconfig.get("current-context", None)
+        if current_context is None:
+            return "error", ErrorOutput(
+                """The provided kubeconfig file does not have a current-context set."""
```
Comment (portante) [inline kubeconfig_plugin.py]: What about these strings requires the triple quotes?

File: kubeconfig_plugin.py [hunk 19]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
+                return "error", ErrorOutput(
+                    "'name' section missing from context entry in kubeconfig"
+                )
+            if ctx["name"] == current_context:
+                context = ctx["context"]
+        if context is None:
             return "error", ErrorOutput(
-                1, "Both client data and token missing, exiting!"
+                "Failed to find a context named {} in the kubeconfig file.".format(
+                    current_context
```
Comment (portante) [inline kubeconfig_plugin.py]: Is there a reason you can't use an f-string here?

File: kubeconfig_plugin.py [hunk 20]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
+                )
             )
 
-        print(output)
+        for expected_key in ["cluster", "user"]:
+            if expected_key not in context:
+                return "error", ErrorOutput(
+                    "'{}' field missing from kubeconfig current context.".format(
```
Comment (portante) [inline kubeconfig_plugin.py]: Perhaps an f-string?

File: kubeconfig_plugin.py [hunk 21]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
+        for cl in kubeconfig["clusters"]:
+            if "name" not in cl:
+                return "error", ErrorOutput(
+                    "'name' section missing from clusters entry in kubeconfig"
+                )
+            if cl["name"] == current_cluster:
+                if "cluster" not in cl:
+                    return "error", ErrorOutput(
+                        "cluster section missing from section "
+                        + "of current cluster {}".format(current_cluster)
```
Comment (portante) [inline kubeconfig_plugin.py]: Python allows you to place string "constants" next to each other to auto combine them.  So something like:

```
"this and" " that"
```

Is equivalent to `"this and that"`.  This also applies across line boundaries, and between f-strings and non-f-strings.  So you can drop the plus sign do just do:

```
"cluster section missing from section of"
f" current cluster {current_cluster}"
```

File: kubeconfig_plugin.py [hunk 22]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
+                    )
+                cluster = cl["cluster"]
+        if cluster is None:
+            return "error", ErrorOutput(
+                "Failed to find a cluster named {} in the kubeconfig file.".format(
```
Comment (portante) [inline kubeconfig_plugin.py]: f-string?

File: kubeconfig_plugin.py [hunk 23]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
+
+        # Now find the user for current context's user for authentication.
+        if "users" not in kubeconfig:
+            return "error", ErrorOutput("'users' section not found in kubeconfig")
+        user = None
+        for u in kubeconfig["users"]:
+            for required_key in ["name", "user"]:
+                if required_key not in u:
+                    return "error", ErrorOutput(
+                        """'{0}' section in the users section not found in the """
```
Comment (portante) [inline kubeconfig_plugin.py]: What requires the triple quotes?

And can we use an f-string here?

File: kubeconfig_plugin.py [hunk 24]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
+                        """kubeconfig""".format(required_key)
+                    )
+            if u["name"] == current_user:
+                user = u["user"]
+        if user is None:
+            return "error", ErrorOutput(
+                "Failed to find a user named {} in the kubeconfig file.".format(
```
Comment (portante) [inline kubeconfig_plugin.py]: f-string?

File: kubeconfig_plugin.py [hunk 25]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
+        output.connection.password = user.get("password", None)
+        output.connection.bearerToken = user.get("token", None)
+
+        return "success", output
+    except Exception as e:
+        # This is the catch-all case.
+        # The goal is for all other errors to be addressed individually.
+        print(traceback.format_exc())
+        return "error", ErrorOutput(
+            "Failure to parse kubeconfig. Exception: {}".format(e.__str__())
```
Comment (portante) [inline kubeconfig_plugin.py]: f-string?

File: kubeconfig_plugin.py [hunk 26]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
+                "Failed to find a user named {} in the kubeconfig file.".format(
+                    current_user
+                )
+            )
+
+        # Ensure the server is in the kubeconfig
+        if "server" not in cluster:
+            return "error", ErrorOutput(
+                """Failed to find server in cluster kubeconfig file {0}"""
+                """ cluster section""".format(current_cluster)
```
Comment (portante) [inline kubeconfig_plugin.py]: Same comment as above ...

File: kubeconfig_plugin.py [hunk 27]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
+        output.connection.key = base64_decode(user.get("client-key-data", None))
+        output.connection.username = user.get("username", None)
+        output.connection.password = user.get("password", None)
+        output.connection.bearerToken = user.get("token", None)
+
+        return "success", output
+    except Exception as e:
+        # This is the catch-all case.
+        # The goal is for all other errors to be addressed individually.
+        print(traceback.format_exc())
```
Comment (portante) [inline kubeconfig_plugin.py]: Why are we printing a traceback and returning an error?

If we do need to print an error, shouldn't it be on `stderr`?

Wouldn't we want to format the stack trace into the exception string returned?  Will the person using the plug-in be able to find the output printed to stdout separate from that returned by the plug-in?

File: kubeconfig_plugin.py [hunk 28]
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
...
-        print(output)
+        for expected_key in ["cluster", "user"]:
+            if expected_key not in context:
+                return "error", ErrorOutput(
+                    "'{}' field missing from kubeconfig current context.".format(
+                        expected_key
+                    )
+                )
+        current_cluster = context["cluster"]
+        current_user = context["user"]
```
Comment (portante) [inline kubeconfig_plugin.py]: Instead of a whole for-loop, could we consider something a bit simpler which will make it easier to understand:

```
        try:
            current_cluster = context["cluster"]
            current_user = context["user"]
        except KeyError as e:
            return "error", ErrorOutput(
                f"{e} field missing from kubeconfig current context."
            )
```

**Review by portante — DISMISSED** [review]
In combination it seems it's worth to consider these suggested changes for maintainability of the code.

File: kubeconfig_plugin.py (hunk 27 above)
Comment (jaredoconnell) [reply kubeconfig_plugin.py]: Added to the error output.

File: kubeconfig_plugin.py (hunk 21 above)
Comment (jaredoconnell) [reply kubeconfig_plugin.py]: Thanks for the Python tips!
